### PR TITLE
updates to tags and branches documentation

### PR DESCRIPTION
**Pull request by semorrison**

### CODE REVIEW

File: templates/contribute/tags_and_branches.md [hunk 1]
```diff
@@ -58,23 +66,36 @@ Mathlib.
...
   * Once the Lean PR has been merged and published in a nightly release, the Std adaptation PR
     * should have its `lean-toolchain` updated to `leanprover/lean4:nightly-YYYY-MM-DD`
     * may be merged into `nightly-testing` as needed.
   * Once the Std adaptation PR has been approved, it can be merged to `bump/v4.X.0`.
 * It is always allowed to merged `bump/v4.X.0` into `nightly-testing`, but not conversely.
-  (Changes to `bump/v4.X.0` have been reviewed, but changes to `nightly-testing` may not be.)
+  (Changes to `bump/v4.X.0` have been reviewed, but changes to `nightly-testing` may nothave been.)
 * When it is time to update Std to a new Lean release,
   *hopefully* all that is required is to make a new PR
   consisting of squash merging `bump/v4.X.0` to `main`.
```
Comment (mhuisi) [inline templates/contribute/tags_and_branches.md:101]: I can't comment on the title below that is not part of the diff, but it contains a typo (`leanprover-community/mathilb4`).

File: templates/contribute/tags_and_branches.md [hunk 2]
```diff
@@ -58,23 +66,36 @@ Mathlib.
...
-    and also labelled 'depends on core changes'.
+  you will need to make a branch, and later open a PR from that branch.
+  (Note that the steps below happen automatically at Mathlib,
+  but need to be done manually for Std.)
+  * If the change was made in `leanprover/lean4#NNNN`,
+    then this branch should be called `lean-pr-testing-NNNN`.
+  * This branch should be based off the branch `nightly-testing-YYYY-MM-DD`
+    where `YYYY-MM-DD` is the date of the nightly release that your Lean PR is based off
+    (you can check this will `script/most-recent-nightly-tag.sh` in the Lean repository).
+  * If this branch does not yet exist, you will need to wait
```
Comment (mhuisi) [inline templates/contribute/tags_and_branches.md:81]: In the bullet point before this, "this branch" refers to `lean-pr-testing-NNNN`, but here it seems to refer to `nightly-testing-YYYY-MM-DD`. In the next bullet point, it again refers to `lean-pr-testing-NNNN`.

File: templates/contribute/tags_and_branches.md [hunk 3]
```diff
@@ -58,23 +66,36 @@ Mathlib.
...
+    making the required adaptations.
+  * When opening the PR, remember to set the base branch to `nightly-testing-YYYY-MM-DD`.
+  * Note that the PR, when approved, will not actually be merged into `nightly-testing-YYYY-MM-DD`
+    but rather into the `bump/v4.X.0` branch.
+  * Please label the PR with the `v4.X.0` and 'depends on core changes' labels.
+    (Or ask for this to be done if you don't have write access.)
   * Once the Lean PR has been merged and published in a nightly release, the Std adaptation PR
     * should have its `lean-toolchain` updated to `leanprover/lean4:nightly-YYYY-MM-DD`
     * may be merged into `nightly-testing` as needed.
   * Once the Std adaptation PR has been approved, it can be merged to `bump/v4.X.0`.
```
Comment (mhuisi) [inline templates/contribute/tags_and_branches.md:96]: Is this the responsibility of the PR author, or an Std maintainer?

File: templates/contribute/tags_and_branches.md [hunk 4]
```diff
@@ -58,23 +66,36 @@ Mathlib.
...
+    (you can check this will `script/most-recent-nightly-tag.sh` in the Lean repository).
+  * If this branch does not yet exist, you will need to wait
+    (and possibly move forward to a subsequent nightly).
+    Contact @semorrison for assistance if needed.
+  * Ideally you will push this branch to the main Std repository; we can provide write access if needed.
+  * The `lean-toolchain` on this branch must contain `leanprover/lean4-pr-releases:pr-release-NNNN`.
+  * You may open a PR from the `lean-pr-testing-NNNN` branch, either before or after
+    making the required adaptations.
+  * When opening the PR, remember to set the base branch to `nightly-testing-YYYY-MM-DD`.
+  * Note that the PR, when approved, will not actually be merged into `nightly-testing-YYYY-MM-DD`
```
Comment (mhuisi) [inline templates/contribute/tags_and_branches.md:89]: Is this important for the person submitting the PR? I.e. is there any action where this note would influence what they should do?

File: templates/contribute/tags_and_branches.md [hunk 5]
```diff
@@ -58,23 +66,36 @@ Mathlib.
...
+    but rather into the `bump/v4.X.0` branch.
+  * Please label the PR with the `v4.X.0` and 'depends on core changes' labels.
+    (Or ask for this to be done if you don't have write access.)
   * Once the Lean PR has been merged and published in a nightly release, the Std adaptation PR
     * should have its `lean-toolchain` updated to `leanprover/lean4:nightly-YYYY-MM-DD`
     * may be merged into `nightly-testing` as needed.
   * Once the Std adaptation PR has been approved, it can be merged to `bump/v4.X.0`.
 * It is always allowed to merged `bump/v4.X.0` into `nightly-testing`, but not conversely.
```
Comment (mhuisi) [inline templates/contribute/tags_and_branches.md:97]: s/merged/merge

File: templates/contribute/tags_and_branches.md [hunk 6]
```diff
@@ -9,43 +9,51 @@ Mathlib.
...
+  This will enable Mathlib's `.olean` cache to include your pull request.
 
 ## Tags and branches
 
 ### `leanprover/lean4`
 
 * Development occurs on the `master` branch.
 * Stable releases and release candidates have tags, e.g. `v4.2.0` or `v4.3.0-rc1`.
-  * To use these, your `lean-toolchain` should contain e.g. `leanprover/lean4:v4.2.0`.
+  * To use these release in a project, your `lean-toolchain` should contain e.g. `leanprover/lean4:v4.2.0`.
```
Comment (mhuisi) [inline templates/contribute/tags_and_branches.md:25]: s/release/releases

File: templates/contribute/tags_and_branches.md [hunk 7]
```diff
@@ -9,43 +9,51 @@ Mathlib.
...
+* Every PR automatically receives a toolchain after CI completes successfully.
+  To use PR #NNNN in a project, your `lean-toolchain` should contain
   `leanprover/lean4-pr-releases:pr-release-NNNN`.
+* Every PR automatically generates a `lean-pr-testing-NNNN` branch at Mathlib,
+  described in detail below, and results from this branch are reported via comments
+  in the PR discussion.
+* For any PR that potentially affects Std or Mathlib, you should base your PR off the HEAD of the `nightly` branch.
+  The combined CI for Std and Mathlib will not run automatically unless this is the case.
+* You can determine the most recent nightly release in the history of your PR by running
+  `script/most-recent-nightly-tag.sh` in the Lean repository.
```
Comment (mhuisi) [inline templates/contribute/tags_and_branches.md:45]: Running this on my machine, I get:
```
mhuisi@DESKTOP-SES5827 ~/lean4 (mhuisi/import-auto-completion)> ./script/most-recent-nightly-tag.sh
fatal: 'nightly' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
2023-11-20
```
2023-11-20 obviously also does not exist, and I believe my branch is based off of the 2023-11-21 nightly.

File: templates/contribute/tags_and_branches.md [hunk 8]
```diff
@@ -87,7 +108,10 @@ Mathlib.
     * This is required for Mathlib's `.olean` caching mechanism.
```
Comment (mhuisi) [inline templates/contribute/tags_and_branches.md:108]: Comment for the line above this: Now that we suggest to open branches in Std itself, "PRs to Mathlib should be made from branches of the Mathlib repository itself" is not different from Std anymore.

File: templates/contribute/tags_and_branches.md (hunk 2 above)
Comment (semorrison) [reply templates/contribute/tags_and_branches.md:81]: Thank you. I am breaking my own rule to never use the words "this" or "it". :-)

File: templates/contribute/tags_and_branches.md (hunk 3 above)
Comment (semorrison) [reply templates/contribute/tags_and_branches.md:96]: Clarified: maintainer.

File: templates/contribute/tags_and_branches.md (hunk 4 above)
Comment (semorrison) [reply templates/contribute/tags_and_branches.md:89]: I've removed this, and made sure this is clearly explained when talking about what a maintainer will do.

File: templates/contribute/tags_and_branches.md (hunk 8 above)
Comment (semorrison) [reply templates/contribute/tags_and_branches.md:108]: Ah, it is: *all* PRs to mathlib should be made in the main repo. It's only `lean-pr-testing-NNNN` branches at Std that I'd like to be on the main repo. I've added "All".

File: templates/contribute/tags_and_branches.md (hunk 7 above)
Comment (semorrison) [reply templates/contribute/tags_and_branches.md:45]: Ooof! Fixed in https://github.com/leanprover/lean4/pull/2947

**Pull request merged**